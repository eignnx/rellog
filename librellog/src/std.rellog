[sig [prefix][suffix][compound]][help
    """
    Relates a list `Compound` to some partitioning of itself into a `Prefix` 
    and a `Suffix`. Also works for text strings.
    """
]
[prefix {}][Suffix][compound Suffix]
[prefix {A ..As}][Suffix][compound {A ..Compound}]
    - [prefix As][suffix Suffix][Compound]

[[comment
    """
    [Prefix][Suffix][Compound]
        - [txt_prefix Prefix][txt_suffix Suffix][txt_compound Compound]
    """
]]

[sig [forwards][backwards]][help "Relates a list to its reversal."]
[forwards {}][backwards {}]
[forwards {A ..As}][Backwards]
    - [forwards As][backwards AsBackwards]
    - [prefix AsBackwards][suffix {A}][compound Backwards]

[[todo
    """
    `[[mode [forwards out][backwards in]]]` loops forever.
    Does it need to be an intrinsic?
    """
]]

[list {Member ..Rest}][Member]
[list {Other ..Rest}][Member]
    - [list Rest][Member]

[sig [list][members]][help "Relates a list to a (sub?)list of its members."]
[List][members {}]
[List][members {Member ..Rest}]
    - [List][members Rest]
    - [List][Member]

[must_be [list {}]]
[must_be [list {X ..Xs}]]
    - [must_be [list Xs]]

[can_be [list {}]]
[can_be [list {X ..Xs}]]
[can_be [list L]]
    - [is_var L]

[can_be [int N]]
    | [is_num N]
    | [is_var N]

[must_be [int N]]
    - [is_num N]

[[meta semidet]]
[sig [list][last]][help "Relates a list to its last element."]
[list {Last}][Last]
[list {X ..Xs}][Last]
    - [list Xs][Last]

[[todo "mode [list out][length in] is broken"]]
[sig [list][length]][help "Relates a list to it's length."]
[list {}][length 0]
[list {X ..Xs}][Length]
    - [list Xs][length XsLength]
    - [pred XsLength][succ Length]


[[meta impure]]
[sig [io {println}]][help "Prints a newline character to stdout."]
[io {println}]
    - [io_writeln ""][stream stdout]

[[meta impure]]
[sig [io {println Text}]][help "Prints `Text` to stdout, followed by a newline character."]
[io {println Text}]
    - [io_writeln Text][stream stdout]

[[meta impure]]
[sig [io {print Text}]][help "Prints `Text` to stdout."]
[io {print Text}]
    - [io_write Text][stream stdout]


[[meta impure]]
[help Signature]
    - [sig Signature][help HelpTxt]
    - [term Signature][text SigTxt]
    - [io {println}]
    - [io {print "Relation: "}]
    - [io {println SigTxt}]
    - [io {println "--------------"}]
    - [io {println HelpTxt}]
    - [io {println}]

[fn {factorial 0}][ret 1]
[fn {factorial 1}][ret 1]
[fn {factorial N}][Ret]
    - [gt N][lt 1]
    - [succ N][pred PredN]
    - [fn {factorial PredN}][ret Fact]
    - [product Ret][x N][y Fact]

[sig [formula][result]][help
    """
    Computes an arithmetic expression given by `Formula` and stores the result 
    in `Result`.


    Available operations are:

        - `{X add Y}` for addition

        - `{X sub Y}` for subtraction

        - `{X mul Y}` for multiplication

        - `{X div Y}` for integer division

        - `{X mod Y}` for euclidean modulus

        - `{neg X}` for negation
    """
]
[formula N][result N]
    - [is_num N]
[formula {A0 mul B0}][result C]
    - [formula A0][result A]
    - [formula B0][result B]
    - [den A][quo B][rem 0][num C]
[formula {A0 div B0}][result C]
    - [formula A0][result A]
    - [formula B0][result B]
    - [num A][den B][quo C][rem Ignore]
[formula {A0 mod B0}][result C]
    - [formula A0][result A]
    - [formula B0][result B]
    - [num A][den B][quo Ignore][rem C]
[formula {A0 add B0}][result C]
    - [formula A0][result A]
    - [formula B0][result B]
    - [subtrahend A][difference B][minuend C]
[formula {A0 sub B0}][result C]
    - [formula A0][result A]
    - [formula B0][result B]
    - [minuend A][subtrahend A][difference C]
[formula {neg A0}][result B]
    - [formula A0][result A]
    - [minuend 0][subtrahend A][difference B]


[[comment "Documentation for builtins"]]

[sig [rel][attrs]][help
    """
    Relates a relation to a list of its attributes. An attribute is a 
    `[key Value]` pair.
    """
]

[sig [attr][key][value]][help
    "Relates an attribute (a `[key Value]` pair) to its key and value."
]

[sig [rel][key][value]][help
    """
    Extracts a value from a relation given the key name. Faster than searching 
    through a list of the relation's attributes.
    """
]

[sig [gt][lt]][help
    "True if the number `Gt` is greater than the number `Lt`."
]

[sig [gte][lte]][help
    "True if the number `Gte` is greater than or equal to the number `Lte`."
]

[[meta non_monotonic]]
[[meta impure]]
[sig [is_var]][help
    "Succeeds if its argument is an unbound variable."
]

[sig [txt_prefix][txt_suffix][txt_compound]][help
    "Pushes a suffix onto a text string."
]

[sig [pred][succ]][help
    "Relates two adjacent integers: a predecessor and a successor."
]

[sig [true]][help "Always succeeds."]
[sig [false]][help "Always fails."]

[[meta impure]]
[sig [io_writeln][stream]][help
    "Writes a text string to `Stream`, and then writes a newline character."
]

[[meta impure]]
[sig [io_write][stream]][help
    "Writes a text string to `Stream`."
]

[[meta non_monotonic]]
[sig [term][text]][help
    "Relates a term to its text string representation."
]

[sig [builtins]][help "Lists all builtin relations."]

[[meta impure]]
[sig [cwd]][help "Gives the current working directory."]

[[meta impure]]
[sig [cd]][help
    """
    Changes the current working directory to the path provided as an argument.
    """
]

[[meta non_monotonic]]
[[meta impure]]
[sig [ls]][help
    """
    Unifies its argument with a list of file/directory names from the current 
    working directory.
    """
]

[[alias_for [minuend][subtrahend][difference]]]
[Min][Sub][Dif]
    - [minuend Min][subtrahend Sub][difference Dif]

[sig [minuend][subtrahend][difference]][help
    """
    Expresses the relationship: `Minuend - Subtrahend = Difference`.
    
    To perform addition, use mode `[subtrahend][difference][Minuend]`.

    Alias: [min][sub][dif]
    """
]

[[alias_for [numerator][denominator][quotient][remainder]]]
[Num][Den][Quo][Rem]
    - [numerator Num][denominator Den][quotient Quo][remainder Rem]

[sig [numerator][denominator][quotient][remainder]][help
    """
    Expresses the relationship: `Numerator/Denominator = Quotient (with a 
    remainder of Remainder)`. More precisely, the relationship is `Numerator = 
    Quotient * Denominator + Remainder`.
    
    Can be used to express divisibility (when Remainder is set to 0), as well as
    multiplication (when Remainder is 0).

    Alias: [num][den][quo][rem]


    Common Modes:

    [num][den][Quo][Rem]   <=> quo = num / den, rem = num % den

    [Num][den][quo][rem]   <=> num = den * quo + rem

    [Num][den 1][quo][rem] <=> num = quo + rem

    [Num][den][quo 1][rem] <=> num = den + rem

    [Num][den][quo][rem 0] <=> num = den * quo

    [num][Den][quo][rem]   <=> den = (num - rem) / quo

    [num][den][Quo][rem]   <=> quo = (num - rem) / den

    [num][Den][quo 1][rem] <=> den = num - rem

    [num][den 1][Quo][rem] <=> quo = num - rem
    """
]

[Nat][Base][Digits]
    - [Nat][Base][RevDigits]
    - [forwards RevDigits][backwards Digits]

[nat 0][Base][revdigits {}]
[Nat][Base][revdigits {Rem ..Ds}]
    - [num Nat][den Base][Quo][Rem]
    - [nat Quo][Base][revdigits Ds]

[[todo
    "[[mode [num out][den in][quo out][rem in]]] not supported."
]]
