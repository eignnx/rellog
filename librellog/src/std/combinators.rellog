[[comment "`c8r` abbreviates 'combinator.'"]]
[c8r {i X}][ret {X}]
[c8r {k X Y}][ret {X}]
[c8r {b F G X}][ret {F {G X}}]
[c8r {c F G X}][ret {{F X} G}]
[c8r {s F G X}][ret {F X {G X}}]
[c8r {m X}][ret {X X}]
[c8r {y F}][ret {F {y F}}]
[c8r {iota X}][ret {X s k}]


[sig [program][val]][help
    """
    Input a program like `[program {iota i 1 2}][Val]`, then backtrack to see
    the reduction steps.
    """
]
[program {X}][val X]
    | [must_be_num X]
    | [must_be_sym X]
[program {{A..As}..Bs}][Val]
    - [prefix {A..As}][suffix Bs][compound ProgramNext]
    - [program ProgramNext][Val]
[Program][Val]
    - [c8r Pat][Ret]
    - [compound Program][prefix Pat][suffix ProgramRem]
    - [prefix Ret][suffix ProgramRem][compound ProgramNext]
    - [program ProgramNext][Val]
[Program][val Program]