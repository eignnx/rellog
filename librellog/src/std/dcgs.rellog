[Rule][Output]
    - [Rule][before ""][after Output]


[Rule][Before][After]
    - [must_be_txt Rule]
    - [txt_prefix Before][txt_suffix Rule][txt_compound After]

[Rule][Before][After]
    - [must_be_rel Rule]
    - [clause_head Rule][ClauseBody]
    - [rule ClauseBody][Before][After]

[Rule][Before][After]
    - [block Rule][functor '-'][Members]
    - [rules Members][Before][After]


[rules {}][before S][after S]
[rules {Rule ..Rules}][Before][After]
    - [Rule][Before][after AfterRule]
    - [Rules][before AfterRule][After]


[story]
    - "Once upon a time there was "
    - [a][StartLetter]
    - " "
    - [Animal][StartLetter]
    - " who was very hungry."


[animal aardvark][start_letter vowel] - "aardvark"
[animal bear][start_letter consonant] - "bear"


[a][start_letter vowel] - "an"
[a][start_letter consonant] - "a"
