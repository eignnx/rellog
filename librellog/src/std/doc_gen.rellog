[gen_docs][Before][After]
    Before = "
    # Documentation
    Multiline?
    [{..After}]"

[Tag][Inner]
    - "<"; Tag; ">"
    - Inner
    - "</"; Tag; ">"

[impl][Sig][Doc]
    - "<section>"
    - "<h1>"; [fmt_sig Sig]; "</h1>"
    - "<p>"; Doc; "</p>"
    - "</section>"

[fmt_sig Sig]
    - logic=[term Sig][Text]
    - Text

