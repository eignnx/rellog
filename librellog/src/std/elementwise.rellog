[[examples
    """
    
    -- [elementwise [sqrt {1 2 3}][sq Out]]
    
        - Out = {1 4 9}

    -- [elementwise [base {1 2 3}][exponent [scalar 2]][power Out]]
    
        - Out = {1 4 9}


    ?- YallLambda = {A,B,C}/[X,Y,Z]>>pred(A,B,C,X,Y,Z).
    % Use `@/1` to open a new lambda scope:
    ?- AtLambda = @( pred(A, B, C, @1, @2, @3) ).
    % Or you can remove the outermost parentheses:
    ?- AtLambda = @pred(A, B, C, @1, @2, @3).

    -- RellogLambda = @[a A][b B][c C][d @d][e @e][f @f]

    -- @[base @{1 2 3}][exponent 2][power @Out]
    -- [ewise [base [ewise {1 2 3}]][exponent 2][power [ewise Out]]]
    - [ewise [base [ewise {1 2 3}]][exponent 2][power [ewise Out]]]
    - `[base `{1 2 3}][exponent 2][power `Out]
    - +[base +{1 2 3}][exponent 2][power +Out]
    - *[base *{1 2 3}][exponent 2][power *Out]
    - $[base ${1 2 3}][exponent 2][power $Out]
    - @[base @{1 2 3}][exponent 2][power @Out]
    - #[base #{1 2 3}][exponent 2][power #Out]
    - &[base #{1 2 3}][exponent 2][power #Out]
    - &[base &{1 2 3}][exponent 2][power &Out]
    - #[base #{1 2 3}][exponent 2][power #Out]
    - :[base :{1 2 3}][exponent 2][power :Out]
    - ^[base ^{1 2 3}][exponent 2][power ^Out]
    - ![base !{1 2 3}][exponent 2][power !Out]
    - ([base ({1 2 3})][exponent 2][power (Out)])

    - [each [base @{1 2 3}][exponent 2][power @Out]]
    - [each [base @{1 2 3}][exponent 2][@Power]]
    - [elementwise
          [base @{1 2 3}][exponent 2][@Power]
      ]
    - [filter
          - [gt @Numbers][lt @Bounds]
          - [member @Numbers][list {3 6 9}]
      ][Filtered]

    {
        [list {1 2 3}][var IterationVar$3]
        [list X][var IterationVar$5]
    }

    {
        [pred 1][succ X0]
        [pred 2][succ X1]
        [pred 3][succ X2]
    }

    {
        {
            [ele 1][var IterationVar$3]
            [ele X0][var IterationVar$5]
        }
        {
            [ele 2][var IterationVar$3]
            [ele X1][var IterationVar$5]
        }
        {
            [ele 3][var IterationVar$3]
            [ele X2][var IterationVar$5]
        }
    }
    """
]]

[call Goal]
    - Goal

[list1 {}][list2 {}][zipped {}]
[list1 {A..As}][list2 {B..Bs}][zipped {{A B}..AsBs}]
    - [list1 As][list2 Bs][zipped AsBs]


[[example
    """
    -- [elementwise
           [base [splat {1 2 3}]
           ][exponent [scalar 3]
           ][power [splat Out]]
       ]
    """
]]
[elementwise Goal]
    - [rel Goal][Attrs]
    - [splatted Attrs][Unsplatted][ListsSplatted]
    - [attrs Unsplatted][rel GoalNew]
    - [ListsSplatted][goal GoalNew]


[Goal][lists_splatted {}]

[Goal][lists_splatted {[list {}][Var]}]
[Goal][lists_splatted {[list {X ..Xs}][Var]}]
    - [original Goal][duplicate GoalDup][renaming {Var}][renamed {VarDup}]
    - VarDup = X
    - GoalDup
    - [lists_splatted {[list Xs][Var]}][Goal]

[Goal][lists_splatted {[list {}][var A] [list {}][var B]}]
[Goal][lists_splatted {[list {X ..Xs}][var A] [list {Y ..Ys}][var B]}]
    - [original {A B Goal}][duplicate {X Y GoalDup}]
    - GoalDup
    - [Goal][lists_splatted {[list Xs][var A] [list Ys][var B]}]



[splatted {}][unsplatted {}][lists_splatted {}]

[splatted {Attr0..Attrs0}][unsplatted {Attr..Attrs}][lists_splatted {[var IterationVar][List] ..Xs}]
    - [attr Attr0][Key][value [splat List]]
    - [Attr][Key][value IterationVar]
    - [splatted Attrs0][unsplatted Attrs][lists_splatted Xs]

[splatted {Attr0..Attrs0}][unsplatted {Attr..Attrs}][ListsSplatted]
    - [attr Attr0][Key][value [Scalar]]
    - [Attr][Key][value Scalar]
    - [splatted Attrs0][unsplatted Attrs][ListsSplatted]


[dbg Term]
    - [Term][Text]
    - [io {println Text}]

[old_list {}][new_list {}][subst [Old][New]]
[old_list {Old..OldRem}][new_list {New..NewRem}][subst [Old][New]]
    - [old_list OldRem][new_list NewRem][subst [Old][New]]
[old_list {Unrelated..OldRem}][new_list {Unrelated..NewRem}][subst [Old][New]]
    - [old_list OldRem][new_list NewRem][subst [Old][New]]

