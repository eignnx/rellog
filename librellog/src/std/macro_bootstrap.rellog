[impl][rule MacroHead ~ [macro MacroInfo]][Before][After]
    - MacroInfo = [NextRelDef]
    - [MacroHead][Options][MacroBody]
    - [provide_options [MacroHead][NextRelDef][Options]]
    - [rule MacroBody][Before][After]


[provide_options [NextRelDef][options {}][MacroHead]]
[provide_options [NextRelDef][options {Opt ..Opts}][MacroHead]]
    - [provide_option [NextRelDef][Opt][MacroHead]]
    - [provide_options [NextRelDef][options Opts][MacroHead]]

[provide_option [next_rel_def [ClauseHead][ClauseBody]][opt [decorates [ClauseHead][ClauseBody]]][MacroHead]]
[provide_option [next_rel_def {}][opt [decorates [ClauseHead][ClauseBody]]][MacroHead]]
    - [tm MacroHead][Txt]
    - [io_write "Could not provide option for next_rel_def: "][stream stderr]
    - [io_writeln Txt][stream stderr]
    - [false]


[sig [Doc]][doc
    """
    A macro that allows inline documentation of a relation's signature.
    """
]
[macro_head
    [Doc]
][options {
    [decorates [ClauseHead][clause_body _ThrowAway]]
}][macro_body
    - logic=[rel ClauseHead][Sig]
    - { [fact [Sig][Doc]] }
]


[sig [ClauseDoc]][doc
    """
    A macro that allows inline documentation of one specific clause of a relation.
    """
]
[macro_head
    [ClauseDoc]
][options {
    [decorates [ClauseHead][clause_body _ThrowAway]]
}][macro_body
    { [fact [sig ClauseHead][doc ClauseDoc]] }
]


[macro_head [Comment]][options {}][macro_body {}]


[macro_head [Todo]][options {}][macro_body
    { [fact [Todo]] }
]


[macro_head [Exporting]][options {}][macro_body {}]


[macro_head [Importing]][options {}][macro_body {}]

[macro_head [Pub]][options {
    [decorates [ClauseHead][clause_body _]]
}][macro_body {}]