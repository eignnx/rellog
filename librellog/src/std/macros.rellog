

[Imports]
    - logic=[tm_in Imports][reified_tm_out ImportsReified]
    - [reified_imports Imports ~ ImportsReified][current_path {}]

[reified_imports ModName ~ [sym _]][CurrentPath]
    - logic=
        - [term ModName][text ModName.text]
        - [Cwd]
        - [txt_prefix Cwd][txt_suffix "/"][txt_compound BaseDir]
        - [txt_prefix BaseDir][txt_suffix ModName.text][txt_compound FilePath]
        - [FilePath][Clauses][directives _]
        - [Clauses][mod_prefix ModName][ClausesPrefixed]
    - ClausesPrefixed


[mod_prefix _][clauses {}][clauses_prefixed {}]

[ModPrefix][clauses {Clause ..Clauses}][clauses_prefixed {ModPrefix::Clause ..ClausesPrefixed}]
    [ModPrefix][Clauses][ClausesPrefixed]


[reified_imports {} ~ _] {}
[reified_imports {Path ..Paths} ~ _]
    - logic=
        - [term ModName][text ModName.text]
        - [Cwd]
        - [txt_prefix Cwd][txt_suffix ModName.text][txt_compound FilePath]
        - [FilePath][Clauses][directives _]
        - [Clauses][mod_prefix ModName][ClausesPrefixed]
    - ClausesPrefixed
