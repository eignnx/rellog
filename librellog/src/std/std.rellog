[sig [Prefix][Suffix][Compound]][doc
    """
    Relates a list `Compound` to some partitioning of itself into a `Prefix` 
    and a `Suffix`. Also works for text strings.
    """
]
[prefix {}][Suffix][compound Suffix]
[prefix {A ..As}][Suffix][compound {A ..Compound}]
    - [prefix As][suffix Suffix][Compound]

[[comment
    """
    [Prefix][Suffix][Compound]
        - [txt_prefix Prefix][txt_suffix Suffix][txt_compound Compound]
    """
]]

[sig [Forward][Backward]][doc "Relates a list to its reversal."]
[forward {}][backward {}]
[forward {A ..As}][Backward]
    - [forward As][backward AsBackward]
    - [prefix AsBackward][suffix {A}][compound Backward]

[[todo
    """
    `[[mode [forward out][backward in]]]` loops forever.
    Does it need to be an intrinsic?
    """
]]

[list {Member ..Rest}][Member]
[list {Other ..Rest}][Member]
    - [list Rest][Member]

[sig [List][Members]][doc "Relates a list to a (sub?)list of its members."]
[List][members {}]
[List][members {Member ..Rest}]
    - [List][members Rest]
    - [List][Member]

[must_be [list {}]]
[must_be [list {X ..Xs}]]
    - [must_be [list Xs]]

[can_be [list {}]]
[can_be [list {X ..Xs}]]
[can_be [list L]]
    - [must_be_var L]

[can_be [int N]]
    | [must_be_num N]
    | [must_be_var N]

[must_be [num N]]
    - [must_be_num N]

[sig [List][Last]][doc "Relates a list to its last element."]
[list {Last}][Last]
[list {X ..Xs}][Last]
    - [list Xs][Last]

[[todo "mode [list out][len in] is broken"]]
[sig [List][Len]][doc "Relates a list to it's length."]
[list {}][len 0]
[list {X ..Xs}][Len]
    - [list Xs][len XsLen]
    - [pred XsLen][succ Len]


[sig [io {println}]][doc "Prints a newline character to stdout."]
[io {println}]
    - [io_writeln ""][stream stdout]

[sig [io {println Text}]][doc "Prints `Text` to stdout, followed by a newline character."]
[io {println Text}]
    - [io_writeln Text][stream stdout]

[sig [io {print Text}]][doc "Prints `Text` to stdout."]
[io {print Text}]
    - [io_write Text][stream stdout]


[sig [io {writeln Text}]][doc "Writes `Text` to `Stream`, followed by a newline character."]
[io {writeln Stream Text}]
    - [io_writeln Text][Stream]

[sig [io {writeln}]][doc "Writes a newline character to `Stream`."]
[io {writeln Stream}]
    - [io_writeln ""][Stream]

[sig [io {write Text}]][doc "Writes `Text` to `Stream`."]
[io {write Stream Text}]
    - [io_write Text][Stream]


[help Signature]
    - [sig Signature][doc DocTxt]
    - [term Signature][text SigTxt]
    - [io {println}]
    - [io {print "Relation: "}]
    - [io {println SigTxt}]
    - [io {println "--------------"}]
    - [io {println DocTxt}]
    - [io {println}]


[[comment "Documentation for builtins"]]

[sig [Rel][Attrs]][doc
    """
    Relates a relation to a list of its attributes. An attribute is a 
    `[key Value]` pair.
    """
]

[sig [Attr][Key][Value]][doc
    "Relates an attribute (a `[key Value]` pair) to its key and value."
]

[sig [Rel][Key][Value]][doc
    """
    Extracts a value from a relation given the key name. Faster than searching 
    through a list of the relation's attributes.
    """
]

[sig [Eq]][doc
    """
    Unifies together the elements of the provided list.
    
    -- [eq {A B C}]
    """
]

[sig [Term][Variables]][doc
    "Relates a term to the list of variables contained in that term."
]

[sig [Original][Duplicate]][doc
    """
    Creates a duplicate of `Original` and unifies it with `Duplicate`.
    """
]

[sig [Original][Duplicate][Renaming][Renamed]][doc
    """
    Creates a duplicate of `Original` and unifies it with `Duplicate`.

    The variables in the list `Renaming` that appear in `Original` will be
    renamed and their new names will be put in the list `Renamed`.

    # Example

    ```
    -- [original [a A][b B][ab {A B}]][Duplicate][renaming {A Z}][Renamed]
        - Duplicate = [a A_1][B][ab {A_1 B}]
        - Renamed = {A_1 Z}
    ```
    """
]

[sig [SnakeCase][PascalCase]][doc
    "Convert symbols between snake_case and PascalCase."
]

[sig [Gt][Lt]][doc
    "True if the number `Gt` is greater than the number `Lt`."
]

[sig [Gte][Lte]][doc
    "True if the number `Gte` is greater than or equal to the number `Lte`."
]

[sig [MustBeVar]][doc
    "Succeeds if its argument is an unbound variable."
]

[sig [MustBeNum]][doc
    "Succeeds if its argument is an instantiated number."
]

[sig [MustBeSym]][doc
    "Succeeds if its argument is an instantiated symbol."
]

[sig [MustBeTxt]][doc
    "Succeeds if its argument is an instantiated text string."
]

[sig [MustBeRel]][doc
    "Succeeds if its argument is an instantiated relation structure."
]

[sig [TxtPrefix][TxtSuffix][TxtCompound]][doc
    "Pushes a suffix onto a text string."
]

[sig [Pred][Succ]][doc
    "Relates two adjacent integers: a predecessor and a successor."
]

[sig [True]][doc "Always succeeds."]
[sig [False]][doc "Always fails."]
[sig [Not]][doc
    """
    Succeeds if the goal passed to it fails, fails if the goal succeeds at least 
    once. Note: This is safe because the knowledgebase is immutable.
    """
]

[sig [IoWriteln][Stream]][doc
    "Writes a text string to `Stream`, and then writes a newline character."
]

[sig [IoWrite][Stream]][doc
    "Writes a text string to `Stream`."
]

[sig [Term][Text]][doc
    "Relates a term to its text string representation."
]

[sig [Builtins]][doc "Lists all builtin relations."]

[sig [Cwd]][doc "Gives the current working directory."]

[sig [Cd]][doc
    """
    Changes the current working directory to the path provided as an argument.
    """
]

[sig [Ls]][doc
    """
    Unifies its argument with a list of file/directory names from the current 
    working directory.
    """
]

[Min][Sub][Dif]
    - [minuend Min][subtrahend Sub][difference Dif]

[sig [Minuend][Subtrahend][Difference]][doc
    """
    Expresses the relationship: `Minuend - Subtrahend = Difference`.
    
    To perform addition, use mode `[subtrahend][difference][Minuend]`.

    Alias: [Min][Sub][Dif]
    """
]

[Num][Den][Quo][Rem]
    - [numerator Num][denominator Den][quotient Quo][remainder Rem]

[sig [Numerator][Denominator][Quotient][Remainder]][doc
    """
    Expresses the relationship: `Numerator/Denominator = Quotient (with a 
    remainder of Remainder)`. More precisely, the relationship is `Numerator = 
    Quotient * Denominator + Remainder`.
    
    Can be used to express divisibility (when Remainder is set to 0), as well as
    multiplication (when Remainder is 0).

    Alias: [Num][Den][Quo][Rem]


    Common Modes:

    [num][den][Quo][Rem]   <=> quo = num / den, rem = num % den

    [Num][den][quo][rem]   <=> num = den * quo + rem

    [Num][den 1][quo][rem] <=> num = quo + rem

    [Num][den][quo 1][rem] <=> num = den + rem

    [Num][den][quo][rem 0] <=> num = den * quo

    [num][Den][quo][rem]   <=> den = (num - rem) / quo

    [num][den][Quo][rem]   <=> quo = (num - rem) / den

    [num][Den][quo 1][rem] <=> den = num - rem

    [num][den 1][Quo][rem] <=> quo = num - rem
    """
]

[sig [FilePath][Clauses][Directives]][doc
    """
    Relates a rellog source file given by `FilePath` to that files contained 
    `Clauses` and `Directives`.
    """
]

[undocumented_sig S]
    - [clause_head S][clause_body B]
    - [not [Sig][Doc] ~ S]
    - [not [Sig][Doc]; (Sig ~ S)]

[sig [Goal][Truth]][doc
    """
    Runs `Goal`. If successful `Truth` = `true`, if it fails `Truth` = `false`,
    and if it errors `Truth` = `[Error]` where `Error` is the error generated.

    When `Goal` errors, all subsequent solutions will be cut.
    """
]

[[comment
    """
    - [tm_in Rule][reified_tm_out ReifiedRule]
    - [impl][rule Rule ~ ReifiedRule][Before][After]
    """
]]

[goal Goal0][MetaParam][new_goal Goal]
    - [rel Goal0][attrs Attrs0]
    - [elementwise [key [scalar MetaParam]][attr_in [splat Attrs0]][attr_out [splat Attrs]]]
    - [rel Goal][Attrs]

[Key][attr_in [key K][value OldValue]][attr_out [key K][value NewValue]]
    [if K = Key][then
        [tm_in OldValue][reified_tm_out NewValue0]; (NewValue = (OldValue ~ NewValue0))
    ][else
        NewValue = OldValue
    ]

[if Cond][Then][Else]
    - [goal Cond][Truth]
    - [rel [true Then][false Else]][key Truth][value Callable]
    - Callable
