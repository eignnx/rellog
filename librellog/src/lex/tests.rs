use insta::{assert_debug_snapshot, glob};
use std::fs;

use crate::{init_interner, utils::my_nom::Span};

#[test]
fn test_tokenize() {
    init_interner();
    glob!("snapshot_tests/*.rellog", |path| {
        let input = fs::read_to_string(path).unwrap();
        let span = Span::new(&input);
        let output = super::lex::tokenize(span, path.into());
        assert_debug_snapshot!(output);
    });
}
