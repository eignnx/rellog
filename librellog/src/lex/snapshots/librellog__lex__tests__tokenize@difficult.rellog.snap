---
source: librellog/src/lex/tests.rs
expression: output
input_file: librellog/src/lex/snapshot_tests/difficult.rellog
---
Ok(
    [
        OBrack@1:1,
        Sym(Sym(provide_option))@1:2,
        OBrack@1:16,
        Sym(Sym(next_rel_def))@1:18,
        OBrace@1:30,
        CBrace@1:32,
        COBrack@1:33,
        Sym(Sym(opt))@1:35,
        OBrack@1:38,
        Sym(Sym(decorates))@1:40,
        OBrack@1:49,
        Var(Var(ClauseHead$))@1:51,
        COBrack@1:61,
        Var(Var(ClauseBody$))@1:63,
        CBrack@1:73,
        CBrack@1:74,
        COBrack@1:75,
        Var(Var(MacroHead$))@1:77,
        CBrack@1:86,
        CBrack@1:87,
        Dash@1:88,
        OBrack@2:6,
        Sym(Sym(tm))@2:8,
        Var(Var(MacroHead$))@2:10,
        COBrack@2:20,
        Var(Var(Txt$))@2:22,
        CBrack@2:25,
        Dash@2:26,
        OBrack@3:6,
        Sym(Sym(io_write))@3:8,
        OQuote@3:16,
        TxtContent("Could not provide option for next_rel_def: ")@3:18,
        CQuote@3:61,
        COBrack@3:62,
        Sym(Sym(stream))@3:64,
        Sym(Sym(stderr))@3:70,
        CBrack@3:77,
        Dash@3:78,
        OBrack@4:6,
        Sym(Sym(io_writeln))@4:8,
        Var(Var(Txt$))@4:18,
        COBrack@4:22,
        Sym(Sym(stream))@4:24,
        Sym(Sym(stderr))@4:30,
        CBrack@4:37,
        Dash@4:38,
        OBrack@5:6,
        Sym(Sym(false))@5:8,
        CBrack@5:13,
    ],
)
